app/scripts/lib/transaction-metrics.test.ts

```diff
@@ -0,0 +1,589 @@
+import {
+  TransactionType,
+  TransactionStatus,
+  AssetType,
+  TokenStandard,
+  TransactionMetaMetricsEvent,
+} from '../../../shared/constants/transaction';
+import {
+  MetaMetricsTransactionEventSource,
+  MetaMetricsEventCategory,
+} from '../../../shared/constants/metametrics';
+import { TRANSACTION_ENVELOPE_TYPE_NAMES } from '../../../shared/lib/transactions-controller-utils';
+import { BlockaidReason } from '../../../shared/constants/security-provider';
+import {
+  onTransactionAdded,
+  onTransactionApproved,
+  onTransactionDropped,
+  onTransactionFinalized,
+  onTransactionRejected,
+  onTransactionSubmitted,
+  METRICS_STATUS_FAILED,
+} from './transaction-metrics';
+///: BEGIN:ONLY_INCLUDE_IN(blockaid)
+///: END:ONLY_INCLUDE_IN
+
+const mockControllerActions = {
+  createEventFragment: jest.fn(),
+  finalizeEventFragment: jest.fn(),
+  getEventFragmentById: jest.fn(),
+  updateEventFragment: jest.fn(),
+  getAccountType: jest.fn(),
+  getDeviceModel: jest.fn(),
+  getEIP1559GasFeeEstimates: jest.fn(),
+  getSelectedAddress: jest.fn(),
+  getTokenStandardAndDetails: jest.fn(),
+  getTransaction: jest.fn(),
+};
+
+const providerResultStub = {
+  eth_getCode: '0x123',
+};
+const { provider } = createTestProviderTools({
+  scaffold: providerResultStub,
+  networkId: '5',
+  chainId: '5',
+});
+
+describe('Transaction metrics', () => {
+  let fromAccount,
+    mockChainId,
+    mockNetworkId,
+    mockTransactionMeta,
+    mockActionId;
+
+  beforeEach(() => {
+    fromAccount = getTestAccounts()[0];
+    mockChainId = '5';
+    mockNetworkId = '5';
+    mockActionId = '2';
+    mockTransactionMeta = {
+      id: '1',
+      status: TransactionStatus.unapproved,
+      txParams: {
+        from: fromAccount.address,
+        to: '0x1678a085c290ebd122dc42cba69373b5953b831d',
+        gasPrice: '0x77359400',
+        gas: '0x7b0d',
+        nonce: '0x4b',
+      },
+      type: TransactionType.simpleSend,
+      origin: ORIGIN_METAMASK,
+      chainId: mockChainId,
+      time: 1624408066355,
+      metamaskNetworkId: mockNetworkId,
+      defaultGasEstimates: {
+        gas: '0x7b0d',
+        gasPrice: '0x77359400',
+      },
+      securityProviderResponse: {
+        flagAsDangerous: 0,
+      },
+      actionId: mockActionId,
+    };
+
+    jest.clearAllMocks();
+  });
+
+  describe('onTransactionAdded', () => {
+    const listener = onTransactionAdded({
+      controllerActions: mockControllerActions,
+      provider: provider as Provider,
+    });
+
+    it('should return if transaction meta is not defined', async () => {
+      await listener({} as any);
+      expect(mockControllerActions.createEventFragment).not.toBeCalled();
+    });
+
+    it('should create event fragment', async () => {
+      await listener({ transactionMeta: mockTransactionMeta as any });
+
+      expect(mockControllerActions.createEventFragment).toBeCalledTimes(1);
+      expect(mockControllerActions.createEventFragment).toBeCalledWith({
```

Comment:
Love that we have specific unit tests for this logic, and that we're asserting all the properties, awesome work! ❤️